src/strategy/rejection.rs

```diff
@@ -0,0 +1,196 @@
+//-
+// Copyright 2017 The proptest developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Provides `Rejection`, a compact representation of why something, such
+//! as a generated value, was rejected.
+
+use std::borrow::Borrow;
+use std::cmp::Ordering;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::rc::Rc;
+
+use self::Rejection::*;
+
+/// The reason for why something, such as a generated value, was rejected.
+///
+/// # Representation
+/// 
+/// This type is currently representationally equivalent to:
+///
+/// ```rust
+/// use std::rc::Rc;
+/// 
+/// pub enum Rejection {
+///     Borrowed(&'static str),
+///     BoxOwned(Box<str>),
+///     RcOwned(Rc<str>),
+/// }
+///
+/// # fn main() {
+/// # use std::mem::size_of;
+/// assert_eq!(size_of::<Rejection>(), 3 * size_of::<usize>());
+/// # }
+/// ```
+///
+/// This is a compact and essentially copy-on-write representation while also
+/// allowing you not to allocate if the rejection message is not dynamic in
+/// nature. This is most efficient and should be used when possible. If you
+/// are allocating and then cloning the allocated `Rejection`, then you should
+/// prefer the `Rejection::RcOwned` variant as it will only bump the reference
+/// count and not actually do any new heap allocation.
+pub enum Rejection {
+    /// The borrowed representation - this is the most efficient if you for
+    /// example have a string in the data section of your program.
+    Borrowed(&'static str),
+    /// A reference counted shared string, use this if you have dynamically
+    /// generated content that is used more than once.
+    RcOwned(Rc<str>),
```

Comment:
If this case weren't here, `Rejection` could be replaced by `Cow<'static, str>`. I don't think the cost of allocating upon every rejection in what's currently the `Rc` case is really worth having a unique smart pointer type.

Reply:
I first considered using [`Bow`](https://docs.rs/bow/1.0.3/bow/enum.Bow.html) since `assert_eq!(32, size_of::<Cow<'static, str>>());` while `assert_eq!(24, size_of::<Bow<'static, str>>());`. But `Filter` does a lot of `.clone()` on `whence` so I included `Rc` as well. With `&'static str` we can avoid allocating at all. What's the downside besides compile times when the interface doesn't become harder to use?